src/main/java/mods/railcraft/common/blocks/tracks/TrackSwitchBase.java

```diff
@@ -184,24 +217,60 @@ public void updateEntity() {
             return;
 
         boolean wasLocked = locked == 0;
+        boolean springState = sprung == 0;
+
+        // Adding carts we just found to lockingCarts
+        List<UUID> lockcarts = getCartsAtLockEntrance();
+        for (UUID cartUUID : lockcarts) {
+            // Carts could follow a loop track through the switch and back to
+            // the other entrance so we remove it from the other list
+            if (springingCarts.contains(cartUUID))
+                springingCarts.remove(cartUUID);
+            lockingCarts.add(cartUUID);
+        }
+
+        // Same as last section except reversed
+        List<UUID> springcarts = getCartsAtSpringEntrance();
+        for (UUID cartUUID : springcarts) {
+            if (lockingCarts.contains(cartUUID))
+                lockingCarts.remove(cartUUID);
+            springingCarts.add(cartUUID);
+        }
+
+        // We only set sprung/locked when a cart enters our track, this is
+        // mainly for visual purposes as the subclass's getBasicRailMetadata()
+        // determines which direction the carts actually take.
+        List<UUID> cartsOnTrack = CartUtils.getMinecartUUIDsAt(
+                getWorld(), tileEntity.xCoord, tileEntity.yCoord,
+                tileEntity.zCoord, 0.3f);
+        for (UUID cartOnTrack : cartsOnTrack) {
+            if (springingCarts.contains(cartOnTrack)) {
+                sprung = SPRING_DURATION;
+                locked = 0;
+                break;
+            } else if (lockingCarts.contains(cartOnTrack)) {
+                locked = SPRING_DURATION;
+                sprung = 0;
+                break;
+            }
+        }
+
         if (locked > 0)
             locked--;
-        if (shouldLockSwitch())
-            locked = SPRING_DURATION;
-
-        boolean springState = sprung == 0;
         if (sprung > 0)
             sprung--;
```

Comment:
Technically, these counter decrements should probably be before the cart searches, else you are shortening the duration by a tick.